x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/TumblingWindow.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.eql.execution.assembler;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.xpack.eql.execution.search.Ordinal;
+import org.elasticsearch.xpack.eql.execution.search.QueryClient;
+import org.elasticsearch.xpack.eql.execution.sequence.SequenceKey;
+import org.elasticsearch.xpack.eql.session.Payload;
+import org.elasticsearch.xpack.eql.util.ReversedIterator;
+
+import java.util.Iterator;
+import java.util.List;
+
+import static org.elasticsearch.action.ActionListener.wrap;
+
+/**
+ * Time-based window encapsulating query creation and advancement.
+ * Since queries can return different number of results, to avoid creating incorrect sequences,
+ * all searches are 'boxed' to a base query.
+ * The base query is initially the first query - when no results are found, the next query gets promoted.
+ * 
+ * This allows the window to find any follow-up results even if they are found outside the initial window
+ * of a base query.
+ */
+public class TumblingWindow implements Executable {
```

Comment:
This is the querying/pagination part from SequenceRuntime; in this version it introduces two new things:
- range queries to the current window (instead of being open ended which made it hard to reason on whether a sequence should be matched or data be discarded)
- pagination/advancement. If there's no more data in the current query, keep looking in the other queries (as long as they have something to match against).